core/base/std_extensions.hpp

```diff
@@ -70,6 +71,17 @@ template <bool B, typename T, typename F>
 using conditional_t = typename std::conditional<B, T, F>::type;
 
 
+template <typename T>
+using decay_t = typename std::decay<T>::type;
+
+
+template <typename T, typename... Args>
```

Comment:
Not sure I understand what I should write here. The entire file already has a comment:

```
 // This header provides implementations of useful utilities introduced into the
 // C++ standard after C++11 (i.e. C++14 and C++17).
 // For documentation about these utilities refer to the newer version of the
 // standard.
```
And there is nothing new here is to say about `make_unique`.  To get the documentation, you just go to [cppreference.com](https://cppreference.com) and type `std::make_unique` into the search bar. You get the documentation page immediately.

Reply:
Two comments though
1. cppreference doesn't say anything about how make_unique has better exception safety and is simpler to use than unique_ptr, you are left to figure it out by yourself which could take some time
2. we do not have the same capabilities as the norm (e.g. the array construction).

I think as I stated, a note that this saves redundant type specification but we lose the unique_ptr capability to specify deleter, among other things, could have its use. Someone looking at this will have to google quite a bit to find why this can even have its interest.

Reply:
> The idea of the header is to provide "missing" functionality, not to document it.

I think we will have to do that though, otherwise anyone which is coming here will have to google everything and be seriously lost. If we add two lines of documentation on those kind of things, not only is it not much overhead but it's also a major help in getting starting into the library. Otherwise, all of this is completely overwhelming, and we can do something to make that feeling pretty much go away.

Reply:
From what I gather here - you are trying to make it easier for people who do not know (advanced) C++ to get started with Ginkgo. I agree we should do something about it, but wouldn't it be better to add a tutorial covering the C++ things we use in Ginkgo to the wiki?

Maybe I'm not getting what kind of documentation you are trying to add. I don't think that anything about the features we have can be explained in 1-2 lines, so they'll have to google for it anyway - what the thing does is quite obvious, as all of these functions/types are one-liners. Why it is useful is a different thing completely, and for that you need at least half-a-page explanation with an example.

Maybe I''ll better understand what you are trying to do if you give me a complete example of what kind of comment you would add.

Reply:
You are right, here is what I was thinking about. Maybe this could be simplified.

```C++
/**
 * make_unique is a utility function used to construct an object of type T and
 * wrap it in a `unique_ptr`, originally introduced in
 * [C++14](https://en.cppreference.com/w/cpp/memory/unique_ptr/make_unique)
 *
 * @internal This function is a preferred helpful replacement for
 * `std::unique_ptr<T>(new T(...))` in as it removes type T redundancy and
 * provides better exception safety.
 */
```

Overall, tutorials are a great thing and should be done, but to me documenting some more advanced C++ tricks/stuff is a complement to that, it allows to have the information available in your editor as you code and have very little chance to miss it. In addition, I think the `@internal` documentation field can be used to specify good programming practice, C++ tricks explanations ..., embedded in the documentation/code, which should help many future and current Ginkgo programmers.

Reply:
Somehow I missed this comment, and only saw it today...

I can sure add the documentation if it will make you happy, but I think it's not helpful, and even counter-productive. I don't like this for several reasons:

1.  Why should `make_unique` be treated in a more special way than all the other C++ functions? If we document `make_unique`, then it feels like it is for some reason more important than all the other functions - and it's not, it is a minor function only used in one unit test (TBH, I regret even adding it, I should have just used `unique_ptr` + `new` in that unit test and avoid a lot of discussion). 
2.  If we document `make_unique` it's only logical that we document all the other functions in `std_extensions.hpp`, but also all the functions from C++(11) we are using. Again, I don't see a reason  why would `make_unique` have to be any more special than anything else. Are you going to document them all? If so, be my guest....
3.  It is not true that `make_unique` is the "preferred" way compared to `unique_ptr<T>(new T)`. There is a serious drawback with it because it needs to have access to the constructor - if the constructor is not public (which is in 99% of the cases in Ginkgo, the only case where this is not true is a unit test, and the constructor is public there only for simplicity), then you cannot use `make_unique`. 
4.  The documentation you propose is incomplete at best, and doesn't explain __any__ of the reasons why to use or not to use `make_unique`. The exception safety is an empty claim here - which will probably be interpreted as "the other solution is not exception safe, so better use this one", which is __not true__ in general. Only in some specific cases where you create multiple temporary unique_ptr objects in a function call, and only if your compiler is not C++17-compliant.  The documentation adds false security to whoever reads it that they understand `make_unique`, while in reality, they don't. If there was no documentation, they would be forced to look up proper documentation and have a better chance of understanding it.

Reply:
BTW, the [libstdc++ manual](https://gcc.gnu.org/onlinedocs/libstdc++/manual/utilities.html) does not mention `unique_ptr` at all, let alone `make_unique`. And the implementation itself with all the comments looks like this:

```c++
  /// std::make_unique for single objects
  template<typename _Tp, typename... _Args>
    inline typename _MakeUniq<_Tp>::__single_object
    make_unique(_Args&&... __args)
    { return unique_ptr<_Tp>(new _Tp(std::forward<_Args>(__args)...)); }

  /// std::make_unique for arrays of unknown bound
  template<typename _Tp>
    inline typename _MakeUniq<_Tp>::__array
    make_unique(size_t __num)
    { return unique_ptr<_Tp>(new remove_extent_t<_Tp>[__num]()); }

  /// Disable std::make_unique for arrays of known bound
  template<typename _Tp, typename... _Args>
    inline typename _MakeUniq<_Tp>::__invalid_type
    make_unique(_Args&&...) = delete;
```

Is it an oversight in the library? I don't think so. It's just that there are several great user documentations of the standard library, and the only thing that's left to discuss either in the code or in the manual are implementation issues - and there aren't any, since `unique_ptr` is something very straightforward.

Reply:
You removed this altogether, fine.

To finish this discussion on my side, here is my reasoning. 

As a principle, the documentation should be self contained. Otherwise the documentation is useless. If we add a piece of code, even if it's (supposed) to match code available/documented in some other projects, there is no reason not to document it here and save time of many people. It is even more important because we aren't even sure our implementations matches the standard available other code. It's also a publicly accessible utility. And yes, I am in favor of documenting the rest of that std extensions file.

Maybe my documentation proposal was incomplete, I wanted to keep it simple (I know it's for temporaries) and it was merely a proposition example to illustrate what I meant. It's also not false security, because indeed there will be no problem when using that function and you don't have to think about this issue in any case at all.

Reply:
Ok, let's move this into an issue: #56